src/Storages/QueryCache.h

```diff
@@ -0,0 +1,221 @@
+#pragma once
+
+#include <memory>
+#include <condition_variable>
+#include <Common/LRUCache.h>
+
+
+namespace DB
+{
+using QueryCachePtr = std::shared_ptr<QueryCache>;
+
+using Data = std::pair<Block, Chunks>;
+
+struct CacheKey
+{
+    CacheKey(ASTPtr ast_, const Block & header_, const Settings & settings_, std::optional<String> username_)
+        : ast(ast_)
+        , header(header_)
+        , settings(settings_)
+        , username(std::move(username_)) {}
+
+    bool operator==(const CacheKey & other) const
+    {
+        return ast->getTreeHash() == other.ast->getTreeHash()
+               && header.getNamesAndTypesList() == other.header.getNamesAndTypesList()
+               && settings == other.settings
+               && username == other.username;
+    }
+
+    ASTPtr ast;
+    Block header;
+    Settings settings;
+    std::optional<String> username;
+};
+
+struct CacheKeyHasher
+{
+    size_t operator()(const CacheKey & key) const
+    {
+        auto ast_info = key.ast->getTreeHash();
+        auto header_info = std::hash<String>{}(key.header.getNamesAndTypesList().toString());
+        auto settings_info = settingsHash(key.settings);
+        auto username_info = std::hash<std::optional<String>>{}(key.username);
+
+        return ast_info.first + ast_info.second * 9273
+               + header_info * 9273 * 9273
+               + settings_info * 9273 * 9273 * 9273
+               + username_info * 9273 * 9273 * 9273 * 9273;
+    }
+
+    private:
+        static size_t settingsHash(const Settings & settings)
+        {
+            size_t hash = 0;
+            size_t coefficient = 1;
+            for (const auto & setting : settings)
+            {
+                hash += std::hash<String>{}(setting.getValueString()) * coefficient;
+                coefficient *= 53;
+            }
+            return hash;
+        }
+};
+
+struct QueryWeightFunction
+{
+    size_t operator()(const Data & data) const
```

Comment:
Weight of data should be member of data. Because this QueryWeightFunction can be called a lot of times in LRUCache during evictions.